avr-hal-generic/src/ep.rs

```diff
@@ -0,0 +1,247 @@
+
+            fn read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Self::Error> {
+                if bytes.len() + offset as usize > $capacity {
+                    return Err(Self::Error::Bounds);
+                }
+                let len = bytes.len();
+                let mut offset = offset as u16;
+                for i in 0..len {
+                    bytes[i] = unsafe { self.eeprom_get_char(offset) };
+                    offset += 1;
+                }
+                Ok(())
+            }
+            fn capacity(&self) -> usize {
+                $capacity
+            }
+        }
+
+        impl $crate::embedded_storage::nor_flash::NorFlash for $Name {
+            const WRITE_SIZE: usize = 1;
+            const ERASE_SIZE: usize = 1;
+            fn erase(&mut self, from: u32, to: u32) -> Result<(), Self::Error> {
+                if to > $capacity {
+                    return Err(Self::Error::Bounds);
+                }
+
+                $crate::avr_device::interrupt::free(|_cs| {
+                    unsafe {
+                        for i in from..to {
+                            self.eeprom_set_address(i as u16);
+
+                            // Now we know that all bits should be erased.
+                            self.eeprom_set_erase_mode();
+
+                            (*<$ep>::ptr()).eecr.write(|w| w.eepe().set_bit()); // Start Erase-only operation.
+                        }
+                    }
+                });
+
+                Ok(())
+            }
+
+            fn write(&mut self, offset: u32, bytes: &[u8]) -> Result<(), Self::Error> {
+                
+                if bytes.len() + offset as usize > $capacity {
+                    return Err(Self::Error::Bounds);
+                }
+
+                let mut offset = offset as u16;
+                for i in bytes {
+                    $crate::avr_device::interrupt::free(|_cs| {
+                        unsafe { self.eeprom_put_char(offset as u16, *i) };
+                        offset += 1;
+                    });
+                }
+                Ok(())
+            }
+        }
+        // AVR supports multiple writes
+        impl $crate::embedded_storage::nor_flash::MultiwriteNorFlash for $Name {}
+    }; // () => {};
+}
+
+#[macro_export]
+macro_rules! impl_atmega_eeprom {
+    (
+        name: $Name:ident,
+        peripheral: $ep:ty,
+        capacity: $capacity:literal,
+        set_address: |$periph_var:ident, $address:ident| $set_address:block,
+
+    ) => {
+        $crate::impl_eeprom_traditional! {
+            name: $Name,
+            peripheral: $ep,
+            capacity: $capacity,
+            set_address: |$periph_var, $address|$set_address,
+            set_erasewrie_mode: |peripheral| {
+                peripheral.eecr.write(|w| {
+                    w.eempe().set_bit(); // Set Master Write Enable bit
+                    w.eepm().val_0x00() // ...and and Erase+Write mode mode..
```

Comment:
We should fix these `val_0x00()` enumerated values in `avr-device` to give them more meaningful names...

Reply:
@Rahix , review fixed